staging/src/k8s.io/client-go/listers/generic_helpers.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package listers
+
+import (
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ResourceIndexer wraps an indexer and resource for a given type.
+// This is intended for use by listers (generated by lister-gen) only.
+type ResourceIndexer[T runtime.Object] struct {
+	indexer  cache.Indexer
+	resource schema.GroupResource
+}
+
+// NamespacedResourceIndexer wraps an indexer, resource, and namespace for a given type.
+// This is intended for use by listers (generated by lister-gen) only.
+type NamespacedResourceIndexer[T runtime.Object] struct {
```

Comment:
Idea:

Why not just have 1 type:

```
ResourceIndexer[T runtime.Object] struct {
	indexer  cache.Indexer
	resource schema.GroupResource
	namespace string // empty for non-namespaced types
}
```

and then just test `l.namespace` for empty.  One code path, one type. Two
constructors functions (`New` and `NewNamespaced`).

Reply:
I thought I had tried that and run into issues, but apparently not; I’ve updated the PR to use a single type.